### ISSUE TITLE
pcap_stats counting packets twice

### ISSUE THREAD

**Issue by guyharris** [the-tcpdump-group/libpcap#69]
Converted from [SourceForge issue 1473764](http://sourceforge.net/support/tracker.php?aid=1473764), submitted by segruts

libpcap 0.9.4 on RH8.

Received packets are counted twice, so invalid
values are returned by pcap_stats().  In pcap-linux.c,
ps_recv is incremented in pcap_read_packet() AND
incrementally added from results of getsockopt
in pcap_stats_linux() in pcap-linux.c

The comments indicate that the value is incremented
in the read because older versions of the linux kernel
may not support the socket option.  Always doing this,
regardless of what the kernel supports yields
inconsistent results.

**Comment by Oppen** [comment]
Related:
`/proc/net/dev:rx_dropped` reflects **both** drops at L1/2 and drops at L3.
However, raw sockets see everything that make it to L2.
As a consequence, packets dropped because of congestion control or (in a case I stumbled upon on the wild) protocols unknown to the kernel, get counted both as received by the socket and as lost.
There's a statistic more relevant, `rx_missed_errors`, but not all drivers implement it.
Luckily, we can use it through `SIOCETHTOOL` when supported, and fall back to `/proc/net/dev` otherwise.

I have half an implementation in my mind, but I wouldn't like to put the effort if it will be rejected in principle. This doesn't mean I won't accept much needed criticism, but if the feature is not wanted at all it doesn't make much sense to implement it.

**Comment by infrastation** [comment]
The original issue was created on 2006-04-20. The increment in `pcap-linux.c:pcap_stats_linux()` is still there:
```c
		handlep->stat.ps_recv += kstats.tp_packets;
		handlep->stat.ps_drop += kstats.tp_drops;
```
However, the only occurrence of `pcap_read_packet` in the current code is in a comment, and tcpdump seems to report the final number of packets (which comes from `ps_recv`) consistently with the number of packets printed.

Is it true that this problem no longer stands?

**Comment by guyharris** [comment]
This problem appears to have been fixed in cbf3dd5b5be3d87fba7608345c231308cd929e90.

And the code path where it occurred was removed when we required, and only supported, memory-mapped packet capture.

**Comment by infrastation** [comment]
Thank you for confirming.
